# Bài thực hành 1.1.3

## (1) Import thư viện, cài đặt cấu hình, lấy và hiển thị dữ liệu
Bước đầu tiên nhằm chuẩn bị môi trường làm việc, bao gồm việc import các thư viện cần thiết cho quá trình xử lý, phân cụm và trực quan hóa dữ liệu, như Pandas, NumPy, Matplotlib, Scikit-learn và Seaborn.


In [2]:
#Nạp thư viện cần thiết
import numpy as np
import pandas as pd
from scipy import stats

In [3]:
#Load dữ liệu
wine = pd.read_csv("red_wine.csv")

In [4]:
#Xem nhanh dữ liệu
display(wine.head())
print("Kích thước dữ liệu:", wine.shape)
wine.info()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

Kích thước dữ liệu: (1599, 12)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


#### Nhận xét:
Tập dữ liệu này là một DataFrame sạch, không có giá trị bị thiếu (1599 entries non-null cho tất cả các cột), bao gồm 1599 hàng và 12 cột. Hầu hết các thuộc tính là các phép đo hóa học và vật lý liên tục của rượu vang, được biểu diễn bằng kiểu dữ liệu float64 (11/12 cột). Biến mục tiêu tiềm năng là quality, có kiểu dữ liệu int64, đại diện cho điểm chất lượng của rượu. Do không có giá trị bị thiếu và dữ liệu đã được định dạng tốt, tập dữ liệu này đã sẵn sàng để tiến hành Phân tích Khám phá Dữ liệu (EDA) và xây dựng các mô hình Học máy, chẳng hạn như mô hình hồi quy hoặc phân loại, để dự đoán chất lượng rượu vang dựa trên các thuộc tính hóa học của nó.


## (2) Tiền xử lý dữ liệu
Tiền xử lý dữ liệu (Data Preprocessing) là quá trình làm sạch, chuyển đổi và chuẩn bị dữ liệu thô để tối ưu hóa cho phân tích hoặc xây dựng mô hình Học máy. Nó giúp giải quyết các vấn đề như giá trị bị thiếu, dữ liệu không nhất quán và nhiễu, đảm bảo dữ liệu đầu vào có chất lượng và độ chính xác cao nhất.


### Kiểm tra giá trị trùng lặp
Kiểm tra trùng lặp (Duplicate Checking) là quá trình xác định và loại bỏ các hàng dữ liệu giống hệt nhau trong tập dữ liệu. Mục đích chính là đảm bảo tính chính xác của phân tích và cải thiện hiệu suất xử lý. Trong phân tích dữ liệu, thường sử dụng các công cụ như Pandas để nhanh chóng tìm kiếm và loại bỏ các bản sao.


In [5]:
# Kiểm tra dữ liệu trống
print("Số giá trị bị thiếu trong mỗi cột:\n", wine.isnull().sum())

# Kiểm tra trùng lặp
before = wine.shape[0]
wine = wine.drop_duplicates()
after = wine.shape[0]
print(f"Đã loại {before - after} dòng trùng lặp.")

Số giá trị bị thiếu trong mỗi cột:
 fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64
Đã loại 240 dòng trùng lặp.


#### Nhận xét:
Tập dữ liệu red_wine_quality chứa 240 bản ghi trùng lặp chính xác. Sự trùng lặp này có thể là do nhiều mẫu rượu vang khác nhau có cùng đặc tính hóa học, hoặc do lỗi nhập liệu. Tuy nhiên, trong tiền xử lý dữ liệu cho Học máy, việc loại bỏ các bản ghi trùng lặp là cần thiết để đảm bảo mỗi tổ hợp thuộc tính chỉ đóng góp một lần vào quá trình huấn luyện mô hình, giúp ngăn ngừa sự thiên lệch và tăng độ tin cậy của kết quả phân tích.


## (3) Thống kê mô tả
Thống kê mô tả (Descriptive Statistics) là quá trình tóm tắt và mô tả các đặc điểm chính của một tập dữ liệu. Nó sử dụng các chỉ số số học như Trung bình (giá trị điển hình), Độ lệch chuẩn (độ phân tán), và các công cụ trực quan để chuyển dữ liệu thô thành thông tin dễ hiểu và có ý nghĩa. Đây là bước đầu tiên để khám phá và hiểu rõ cấu trúc của dữ liệu.


In [6]:
#Thống kê mô tả dữ liệu
desc = wine.describe().T
desc

count       mean        std      min      25%      50%  \
fixed acidity         1359.0   8.310596   1.736990  4.60000   7.1000   7.9000   
volatile acidity      1359.0   0.529478   0.183031  0.12000   0.3900   0.5200   
citric acid           1359.0   0.272333   0.195537  0.00000   0.0900   0.2600   
residual sugar        1359.0   2.523400   1.352314  0.90000   1.9000   2.2000   
chlorides             1359.0   0.088124   0.049377  0.01200   0.0700   0.0790   
free sulfur dioxide   1359.0  15.893304  10.447270  1.00000   7.0000  14.0000   
total sulfur dioxide  1359.0  46.825975  33.408946  6.00000  22.0000  38.0000   
density               1359.0   0.996709   0.001869  0.99007   0.9956   0.9967   
pH                    1359.0   3.309787   0.155036  2.74000   3.2100   3.3100   
sulphates             1359.0   0.658705   0.170667  0.33000   0.5500   0.6200   
alcohol               1359.0  10.432315   1.082065  8.40000   9.5000  10.2000   
quality               1359.0   5.623252   0.823578  3.00000   5.0000   6.0000   

                           75%        max  
fixed acidity          9.20000   15.90000  
volatile acidity       0.64000    1.58000  
citric acid            0.43000    1.00000  
residual sugar         2.60000   15.50000  
chlorides              0.09100    0.61100  
free sulfur dioxide   21.00000   72.00000  
total sulfur dioxide  63.00000  289.00000  
density                0.99782    1.00369  
pH                     3.40000    4.01000  
sulphates              0.73000    2.00000  
alcohol               11.10000   14.90000  
quality                6.00000    8.00000

#### Nhận xét: 
-Tổng quan & Kích thước dữ liệu Count (Số lượng): Tất cả 12 thuộc tính đều có 1359 quan sát. Điều này cho thấy sau quá trình tiền xử lý, tập dữ liệu không còn giá trị thiếu (missing values) hay đã được xử lý các hàng có giá trị thiếu/trùng lặp, đảm bảo tính nhất quán cho việc phân tích.
-Nhận xét về Xu hướng Trung tâm (Mean & 50%)
+pH: Giá trị trung bình mean ~ 3.31 và trung vị 50% ~ 3.31 rất gần nhau, cho thấy phân phối của pH khá đối xứng.
+Alcohol: Trung bình mean ~ 10.43 và trung vị 50~ 10.20 cũng gần nhau.
+Citric Acid: Trung bình mean~0.27 lớn hơn trung vị 50% ~0.26 một chút, cho thấy phân phối hơi lệch phải (positive skew).
+Volatile Acidity & Chlorides: Giá trị trung bình lớn hơn đáng kể so với trung vị, báo hiệu phân phối lệch phải mạnh, có thể do sự hiện diện của các giá trị ngoại lai cao.
-Nhận xét về Độ phân tán (Std & Min/Max)
+Total Sulfur Dioxide (Lớn): Có độ lệch chuẩn(std =33.4) và khoảng biến thiên (range: 6.0 → 289.0) rất lớn, cho thấy các giá trị thuộc tính này phân tán rộng và có khả năng chứa nhiều giá trị ngoại lai (outliers) cao.
+Free Sulfur Dioxide (Lớn): Độ lệch chuẩn (std = 10.4) và khoảng biến thiên (1.0→72.0) cũng lớn, xác nhận sự phân tán rộng rãi.
+Density (Nhỏ): Độ lệch chuẩn (std =  0.0018) rất nhỏ, cho thấy các giá trị cực kỳ tập trung xung quanh giá trị trung bình (mean = 0.997), gần như không có sự khác biệt.


# Bài thực hành 1.1.4

## (1) Import thư viện, cài đặt cấu hình, lấy và hiển thị nhanh dữ liệu


In [7]:
#Nạp thư viện cần thiết
import numpy as np
import pandas as pd
from scipy import stats

In [13]:
#Load dữ liệu
diabete = pd.read_csv("diabetes.csv")

In [14]:
#Xem nhanh dữ liệu
display(diabete.head())
print("Kích thước dữ liệu:", diabete.shape)
diabete.info()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

Kích thước dữ liệu: (768, 9)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


#### Nhận xét:
Kết quả hiển thị nhanh về bộ dữ liệu Pima Indians Diabetes. Bộ dữ liệu Pima Indians Diabetes có kích thước nhỏ gọn với 768 hàng và 9 cột, được thiết lập cho bài toán phân loại nhị phân (biến mục tiêu Outcome là 0 hoặc 1). Về mặt cấu trúc, dữ liệu rất sạch vì không có giá trị thiếu (Non-Null Count là 768 cho tất cả các cột). 


## (2) Tiền xử lý dữ liệu
Mục tiêu nhằm làm sạch, chuyển đổi và chuẩn bị dữ liệu thô để tối ưu hóa cho phân tích hoặc xây dựng mô hình Học máy. Nó giúp giải quyết các vấn đề như giá trị bị thiếu, dữ liệu không nhất quán và nhiễu, đảm bảo dữ liệu đầu vào có chất lượng và độ chính xác cao nhất.


### Kiểm tra dữ liệu trống và xử lý các giá trị trùng lặp
Ta dùng .isnull().sum() để nhanh chóng xác định tổng số giá trị thiếu (NaN) ở mỗi cột, giúp định hướng cách xử lý (ví dụ: điền bằng trung vị). Đồng thời, ta sử dụng .drop_duplicates() để loại bỏ các hàng trùng lặp hoàn toàn, đảm bảo mỗi quan sát là duy nhất, từ đó ngăn ngừa mô hình bị thiên lệch và tăng độ tin cậy của kết quả.



In [15]:
# Kiểm tra dữ liệu trống
print("Số giá trị bị thiếu trong mỗi cột:\n", diabete.isnull().sum())

# Kiểm tra trùng lặp
before = diabete.shape[0]
diabete = diabete.drop_duplicates()
after = diabete.shape[0]
print(f"Đã loại {before - after} dòng trùng lặp.")

Số giá trị bị thiếu trong mỗi cột:
 Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64
Đã loại 0 dòng trùng lặp.


#### Nhận xét:
Kết quả xử lý giá trị trống và trùng lặp
Bộ dữ liệu Pima Indians Diabetes hiện đang ở trạng thái rất sạch và sẵn sàng để xây dựng mô hình. Việc kiểm tra cho thấy 0 giá trị bị thiếu ở tất cả các cột và 0 dòng trùng lặp đã được loại bỏ. Điều này xác nhận rằng các bước tiền xử lý trước đó—bao gồm việc xử lý các giá trị $0$ không hợp lý (được mã hóa thành NaN và sau đó được điền)—đã được hoàn thành, loại bỏ các vấn đề về chất lượng dữ liệu và đảm bảo tính nhất quán cho các thuật toán Học máy.


## (3) Thống kê mô tả
Thống kê mô tả (Descriptive Statistics) là quá trình tóm tắt và mô tả các đặc điểm chính của một tập dữ liệu. Nó sử dụng các chỉ số số học như Trung bình (giá trị điển hình), Độ lệch chuẩn (độ phân tán), và các công cụ trực quan để chuyển dữ liệu thô thành thông tin dễ hiểu và có ý nghĩa. Đây là bước đầu tiên để khám phá và hiểu rõ cấu trúc của dữ liệu.


In [19]:
#Thống kê mô tả dữ liệu
esc = diabete.describe()
esc

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

#### Nhận xét:
Bộ dữ liệu Pima Indians Diabetes là một tập dữ liệu nhỏ gọn với 768 quan sát, được thiết lập cho bài toán phân loại nhị phân Outcome. 
Dữ liệu về sức khỏe: Tập dữ liệu bao gồm nhiều thuộc tính sinh học quan trọng như Glucose,BloodPressure,BMI, vàInsulin, cùng với các yếu tố nhân khẩu học Age, Pregnancies. Điều này tạo cơ sở vững chắc để nghiên cứu mối quan hệ giữa các chỉ số này và kết quả Outcome.
Độ biến thiên cao: Các thuộc tính như Insulin (std = 115.2) và Glucose (std = 31.9) cho thấy sự đa dạng lớn về mặt giá trị trong các quan sát. Sự đa dạng này giúp mô hình Học máy có thể học hỏi từ một phạm vi rộng các trường hợp y tế.
Xu hướng phân phối:Cột Age có khoảng giá trị rộng (21 đến 81) và xu hướng tập trung ở độ tuổi trẻ hơn (median = 29, mean = 33.2).Cột BMI có (mean = 32.0) gần median 32.0), cho thấy một phân phối khá tập trung xung quanh giá trị trung tâm.
Phân bố lớp: Biến mục tiêu Outcome có mean = 0.349, cho thấy tỷ lệ người mắc bệnh là khoảng 34.9%. Tỷ lệ này cung cấp một cơ hội để phát triển một mô hình có khả năng phân biệt giữa hai nhóm với độ chính xác cao.
